### ISSUE TITLE
Add peek_ret

### ISSUE THREAD

**Issue by Grazfather** [hugsy/gef#150]
Duh.

**Comment by Grazfather** [comment]
This one will be more difficult:

* Need `is_ret` in Architecture classes
* Need a way to fetch the target

**Comment by hugsy** [comment]
I gave it some thoughts:

The (only) big difficulty I foresee for this is for x86 binaries compiled with `-fomit-frame-pointer` and ARM based binaries. The other archs (MIPS, PPC) are rather explicit as to how/where to return (usually through registers). For x86, if frame pointer is used, it is easy as well.

However ARM would always require to parse every `pop` instruction and determine based on its operands how many stack pop we should (for example if I wasn't clear, `pop {r7, r8, pc}` will be different from `pop {r7, r8, r9, pc}` and we cannot know that we're popping `pc` without monitoring every `pop` instruction  which would be a huge downgrade of speed on ARM). And almost the same thing applies to x86 bins compiled with `-fomit-frame-pointer` 

I don't see a (hard or easy) way around this, but totally open to discussion :smiley:

**Comment by Grazfather** [comment]
Yeah, but we can do it like we do peek_calls: If it can't figure out the call target, it just doesn't peek.

For ARM you just split the operands, count them, and get that index from the stack (pc will always be last).

that won't work for leaf functions, which have the RA in LR.

**Comment by Grazfather** [comment]
I should have spent 5 minutes looking at the API docs.

`hex(gdb.selected_frame().older().pc())`

**Comment by hugsy** [comment]
Yeah man, that's a feature cheap to implement :smile:

**Comment by Grazfather** [comment]
Not as easy as that, in the top frame, it can't detect where it'd return, even though it returns to something like `__libc_start_main`.

I've added a method `is_ret` and `get_ra`. `get_ra` could be combined with `is_ret`, perhaps.

`get_ra` is arch specific: If it's on a return instruction it tries to figure out where that would be. On ARM it can be tricky: It can be at the top of the stack, in lr, or farther in the stack. I determine where it is based on the instruction itself.

**Comment by Grazfather** [comment]
the peek_ret branch is ready for testing.

Need to test it especially on SPARC, PPC, and MIPS.

**Comment by hugsy** [comment]
No prob, I can do that tonight.
